### ISSUE TITLE
Add script that uploads a dataset labeled by a reward models and computes the BT probabilities.

### ISSUE THREAD

**Issue by mauriciogtec** [vermashresth/RLAIF-KL-Error#10]
TODOs:

The PR should adapt the code below worked by @acwkim, but it should remove hard-coding (e.g., the reward model). Instruction of how to run it should be added to the README.

### Reference code

```python
import os
import sys
from dataclasses import dataclass, field
from tqdm.auto import tqdm
from huggingface_hub import HfApi
import torch
import numpy as np
from transformers import (
    AutoModel,
    AutoTokenizer,
    HfArgumentParser,
)
from transformers import logging
from accelerate import Accelerator
from accelerate.utils import gather_object
from datasets import load_dataset, DatasetDict


sys.path.append(os.path.dirname(os.path.dirname(os.path.abspath(__file__))))
from utils import CONFIGS, sample_every_k_batched
from safe_rlhf.models import AutoModelForScore

HUGGINGFACE_CONFIGS = CONFIGS.services.huggingface
HFAPI = HfApi(HUGGINGFACE_CONFIGS["token"])
REWARD_CONFIGS = CONFIGS.evaluations.reward
CACHE_CONFIGS = CONFIGS.utils.cache

accelerator = Accelerator()
tqdm.pandas()
logging.set_verbosity_error()


@dataclass
class ScriptArguments:
    run_name: str = field(
        metadata={"help": "run name to evaluate"},
    )
    tag: str = field(
        metadata={"help": "tag for the experiment"},
    )
    split: str = field(
        default="default",
        metadata={"help": "dataset split to evaluate"},
    )
    every_k: int = field(
        default=1,
        metadata={
            "help": "evaluate every k samples, if a fraction, evaluate each sample 1/k times"
        },
    )
    per_device_evalreward_batch_size: int = field(
        default=16,
        metadata={"help": "batch size per device"},
    )
    model_max_length: int = field(
        default=2048,
        metadata={
            "help": "maximum sequence length, sequences will be right padded (and possibly truncated)"
        },
    )
    padding: bool = field(
        default=True,
        metadata={"help": "use padding for tokenizer"},
    )
    truncation: bool = field(
        default=True,
        metadata={"help": "allow truncation for tokenizer"},
    )
    score_model_id: str = field(
        default="none",
        metadata={"help": "model id for scoring, do not change"},
    )
    model_cache_dir: str = field(
        default=CACHE_CONFIGS["model_cache_dir"],
        metadata={"help": "model cache directory"},
    )
    dataset_cache_dir: str = field(
        default=CACHE_CONFIGS["dataset_cache_dir"],
        metadata={"help": "dataset cache directory"},
    )


# Load dataset and remove unnecessary columns
def load_generated_dataset(script_args):
    response_dataset = load_dataset(
        HUGGINGFACE_CONFIGS["prefix"]["evaluations"] + script_args.run_name,
        name=script_args.tag,
        cache_dir=script_args.dataset_cache_dir,
    )

    # Get the specified split
    if script_args.split in response_dataset:
        eval_dataset = response_dataset[script_args.split]
    else:
        print(f"Warning: Split '{script_args.split}' not found. Available splits: {list(response_dataset.keys())}")
        # Try to use 'default' split, if not available use the first available split
        if "default" in response_dataset:
            eval_dataset = response_dataset["default"]
        else:
            first_split = list(response_dataset.keys())[0]
            print(f"Using first available split: {first_split}")
            eval_dataset = response_dataset[first_split]
    
    print(f"Evaluation dataset size: {len(eval_dataset)}")
    return eval_dataset


# Load model and tokenizer for inference
def load_score_model(script_args):
    print(f"Loading score model: {script_args.score_model_id}")
    if script_args.score_model_id.startswith("PKU-Alignment"):
        model = AutoModelForScore.from_pretrained(
            script_args.score_model_id,
            torch_dtype=torch.bfloat16,
            device_map={"": accelerator.local_process_index},
            # Update transformers package to >=4.38.0 and no need to use trust_remote_code
            trust_remote_code=False,
            use_cache=True,
            cache_dir=script_args.model_cache_dir,
        )
        tokenizer = AutoTokenizer.from_pretrained(
            script_args.score_model_id,
            model_max_length=script_args.model_max_length,
            padding_side="left",
            use_fast=True,
            cache_dir=script_args.model_cache_dir,
        )
    elif script_args.score_model_id.startswith("openbmb"):
        model = AutoModel.from_pretrained(
            script_args.score_model_id,
            torch_dtype=torch.bfloat16,
            device_map={"": accelerator.local_process_index},
            # This is needed as EurusRewardModel is not in transformers' model registry
            trust_remote_code=True,
            use_cache=True,
            cache_dir=script_args.model_cache_dir,
        )
        tokenizer = AutoTokenizer.from_pretrained(
            script_args.score_model_id,
            model_max_length=script_args.model_max_length,
            padding_side="left",
            use_fast=True,
            cache_dir=script_args.model_cache_dir,
        )
    elif script_args.score_model_id.startswith("OpenAssistant"):
        from transformers import AutoModelForSequenceClassification
        model = AutoModelForSequenceClassification.from_pretrained(
            script_args.score_model_id,
            torch_dtype=torch.bfloat16,
            device_map={"": accelerator.local_process_index},
            cache_dir=script_args.model_cache_dir,
        )
        tokenizer = AutoTokenizer.from_pretrained(
            script_args.score_model_id,
            model_max_length=script_args.model_max_length,
            padding_side="left",
            use_fast=True,
            cache_dir=script_args.model_cache_dir,
        )
    else:
        raise ValueError("No reward model found for the dataset")
    tokenizer.pad_token = tokenizer.eos_token
    return model, tokenizer


# Evaluate reward of responses
def evaluate_reward(model, tokenizer, response_dataset, script_args):
    generator, num_iters = sample_every_k_batched(
        response_dataset,
        script_args.every_k,
        batch_size=script_args.per_device_evalreward_batch_size,
    )
    # Progress bar only on main process
    pbar = tqdm(total=num_iters, disable=not accelerator.is_local_main_process)
    all_indices_and_samples = list(generator)
    with accelerator.split_between_processes(
        all_indices_and_samples
    ) as process_indices_and_samples:
        # Inference in batch
        result = {}
        for indices, samples in process_indices_and_samples:
            with torch.no_grad():
                prompts = [sample["prompt"] for sample in samples]
                chosens = [sample["chosen"] for sample in samples]
                rejecteds = [sample["rejected"] for sample in samples]

                chosen_texts = [p + c for p, c in zip(prompts, chosens)]
                rejected_texts = [p + r for p, r in zip(prompts, rejecteds)]

                chosen_inputs = tokenizer(
                    chosen_texts,
                    max_length=script_args.model_max_length,
                    truncation=script_args.truncation,
                    padding=script_args.padding,
                    return_tensors="pt",
                ).to(model.device)
                rejected_inputs = tokenizer(
                    rejected_texts,
                    max_length=script_args.model_max_length,
                    truncation=script_args.truncation,
                    padding=script_args.padding,
                    return_tensors="pt",
                ).to(model.device)

                chosen_outputs = model(**chosen_inputs)
                rejected_outputs = model(**rejected_inputs)

               
                if script_args.score_model_id.startswith("PKU-Alignment"):
                    chosen_scores = chosen_outputs.end_scores.cpu().flatten().tolist()
                    rejected_scores = rejected_outputs.end_scores.cpu().flatten().tolist()
                elif script_args.score_model_id.startswith("openbmb"):
                    chosen_scores = chosen_outputs.cpu().flatten().tolist()
                    rejected_scores = rejected_outputs.cpu().flatten().tolist()
                elif script_args.score_model_id.startswith("OpenAssistant"):
                    # AutoModelForSequenceClassification returns logits, we need to get the score for the single class
                    # For OpenAssistant, output is SequenceClassifierOutput, use logits
                    chosen_scores = chosen_outputs.logits.cpu().flatten().tolist()  # Flatten all logits
                    rejected_scores = rejected_outputs.logits.cpu().flatten().tolist()  # Flatten all logits

                else:
                    raise RuntimeError("Unknown reward model")

                bt_probs = [float(torch.sigmoid(torch.tensor(cs - rs))) for cs, rs in zip(chosen_scores, rejected_scores)]

                # Store results as before, but now for three columns
                for idx, cs, rs, prob in zip(indices, chosen_scores, rejected_scores, bt_probs):
                    if idx not in result:
                        result[idx] = {"chosen_score": [], "rejected_score": [], "bt_prob": []}
                    result[idx]["chosen_score"].append(cs)
                    result[idx]["rejected_score"].append(rs)
                    result[idx]["bt_prob"].append(prob)

            if accelerator.is_local_main_process:
                # Update progress bar
                pbar.update(accelerator.num_processes)
        # Transform to list of dicts, otherwise gather_object() will not collect correctly
        result = [result]

        gathered = gather_object(result)
        results = {}
        for d in gathered:
            for k, v in d.items():
                if k not in results:
                    results[k] = {"chosen_score": [], "rejected_score": [], "bt_prob": []}
                results[k]["chosen_score"].extend(v["chosen_score"])
                results[k]["rejected_score"].extend(v["rejected_score"])
                results[k]["bt_prob"].extend(v["bt_prob"])

        def get_column(col):
            return [np.mean(results[idx][col]) if idx in results else None for idx in range(len(response_dataset))]

        for col in ["chosen_score", "rejected_score", "bt_prob"]:
            if col in response_dataset.column_names:
                response_dataset = response_dataset.remove_columns(col)
            response_dataset = response_dataset.add_column(col, get_column(col))

        return response_dataset


def main():
    print('Start Evalreward.py!!')
    parser = HfArgumentParser(ScriptArguments)
    script_args = parser.parse_args_into_dataclasses()[0]

    # Use fixed reward model
    script_args.score_model_id = "PKU-Alignment/beaver-7b-v1.0-reward"

    # Model & Tokenizer
    model, tokenizer = load_score_model(script_args)
    model.eval()

    # Dataset
    # response_dataset = load_generated_dataset(script_args)

    # # Evaluation
    # response_dataset = evaluate_reward(model, tokenizer, response_dataset, script_args)

    # # Save locally instead of pushing to Hub
    # # output_dir = f"eval_results_{script_args.tag}"
    # # response_dataset.save_to_disk(output_dir)
    # # print(f"Results saved to local directory: {output_dir}")

    # DatasetDict(
    #     {"default": response_dataset},
    # ).push_to_hub(
    #     HUGGINGFACE_CONFIGS["prefix"]["evaluations"] + script_args.run_name,
    #     script_args.tag,
    # )
    # print('Finished Evalreward.py')
# Load the full DatasetDict (all splits)
    response_dataset_dict = load_dataset(
        HUGGINGFACE_CONFIGS["prefix"]["evaluations"] + script_args.run_name,
        name=script_args.tag,
        cache_dir=script_args.dataset_cache_dir,
    )

    # Evaluate each split and store results
    result_dict = {}
    for split_name, split_dataset in response_dataset_dict.items():
        print(f"Evaluating split: {split_name} (size: {len(split_dataset)})")
        result_dict[split_name] = evaluate_reward(model, tokenizer, split_dataset, script_args)

    # Save and push with all splits preserved
    DatasetDict(result_dict).push_to_hub(
        HUGGINGFACE_CONFIGS["prefix"]["evaluations"] + script_args.run_name,
        script_args.tag,
    )


if __name__ == "__main__":
    main()
```
